NIFI-4927 - InfluxDB Query Processor

mans2singh opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

mans2singh commented:
Hi Nifi Team:

Please let me know your suggestions/recommendations on this InfluxDB Query Processsor.

Thanks

joewitt commented:
@mans2singh thanks for contributing.  What we need most right now is help reviewing outstanding PRs.  This involves reviewing code and validating as possible whether things are actually functioning.  For things like this it can be tougher because it means someone needs an environment with influx running/etc.. and familiarity on how to do so.  What can you do?  You can provide detailed instructions a reviewer could follow to verify functionality.  Perhaps a pointer to a docker container which eases the process.  And help with reviews on other items, for example.  Not saying someone wont look at this until then - they certainly might.  Just giving you tips to help ease the process.

MikeThomsen commented:
@joewitt @mans2singh [Here's a starting point](https://github.com/apache/nifi/pull/2101#issuecomment-361083921) on the Docker/Docker Compose

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
```
Making input optional could be quite useful. See ExecuteSQL and GetMongo for an example of how to support timer and event driving.

mans2singh replied:
Added support for timer based queries based on ExecuteSQL processesor as you had recommended and updated test cases.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+@EventDriven
+@SupportsBatching
+@Tags({"influxdb", "measurement","get", "read", "query", "timeseries"})
+@CapabilityDescription("Processor to execute InfluxDB query from the content of a FlowFile.  Please check details of the supported queries in InfluxDB documentation (https://www.influxdb.com/).")
```
Based on the wording, should a user assume that all influxdb queries are supported or only some?

mans2singh replied:
Influx queries supports data queries, dml and schema exploration.  I've added integration tests for these.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+@CapabilityDescription("Processor to execute InfluxDB query from the content of a FlowFile.  Please check details of the supported queries in InfluxDB documentation (https://www.influxdb.com/).")
+@WritesAttributes({
+    @WritesAttribute(attribute = AbstractInfluxDBProcessor.INFLUX_DB_ERROR_MESSAGE, description = "InfluxDB error message"),
+    @WritesAttribute(attribute = ExecuteInfluxDBQuery.INFLUX_DB_EXECUTED_QUERY, description = "InfluxDB executed query"),
```
We've started making this user-defined in some other processors.

mans2singh replied:
Let me know if you have any recommendation for this.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        session.exportTo(flowFile, baos);
+        String query = new String(baos.toByteArray(), charset);
```
Just to be safe, you might want to close `baos` before calling `toByteArray()`

mans2singh replied:
Corrected it.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+        try {
+            long startTimeMillis = System.currentTimeMillis();
+            QueryResult result = executeQuery(context, database, query, queryResultTimeunit);
+            String json = gson.toJson(result);
```
Can you be sure that this will not be null and do you need to check for that?

mans2singh replied:
If there is problem in executing query it throws an exception or returns true for `hasError` method and I check for them.  But if you think I should check for `null`, I can do that.

MikeThomsen replied:
If you're pretty confident that the client API isn't going to surprise you with a null response, I'm fine with that.

mans2singh replied:
Added check for null and unit test to support it.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+                session.transfer(flowFile, REL_FAILURE);
+            }
+
+            session.getProvenanceReporter().send(flowFile, new StringBuilder("influxdb://")
```
the `StringBuilder` part should be turned into a method call or broken up to make it easier to read, especially from within a debugger.

mans2singh replied:
Refactored the string building to a method.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+                (endTimeMillis - startTimeMillis));
+        } catch (Exception exception) {
+            flowFile = populateErrorAttributes(session, flowFile, query, exception.getMessage());
+            if ( exception.getCause() instanceof SocketTimeoutException ) {
```
Are there any other network exceptions that need to be caught here?

mans2singh replied:
I did not find any documentation on any other network exception on invoking query.  But let me know if you have any recommendations.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,199 @@
+    protected FlowFile populateErrorAttributes(final ProcessSession session, FlowFile flowFile, String query,
+            String message) {
+        Map<String,String> attributes = new HashMap<>();
+        attributes.put(INFLUX_DB_ERROR_MESSAGE, String.valueOf(message));
```
What are you using `String.valueOf` here?

mans2singh replied:
String.valueOf is a null safe method, it returns string `null` in case string is null otherwise the string, so I can safely add it to the Map.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/AbstractITInfluxDB.java:
```diff
@@ -0,0 +1,79 @@
+
+    protected void initInfluxDB() throws InterruptedException, Exception {
+        influxDB = InfluxDBFactory.connect(dbUrl,user,password);
+        if ( influxDB.databaseExists(dbName) ) {
```
I would strongly suggest moving this to the `@After` section because it'll make everything behave in one clean arc of setup -> test -> cleanup.

mans2singh replied:
I've refactored the code and call it both from `Before` and `After` to make sure that the test setup and tear down is clean.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,163 @@
+    @Before
+    public void setUp() throws Exception {
+        runner = TestRunners.newTestRunner(ExecuteInfluxDBQuery.class);
+        initializeRunner();
```
This can be merged into the other init function. It should at least come after the database init code so that you don't spend any time spinning up testing infra on the NiFi side if the database isn't working.

mans2singh replied:
I've changed the order and call init database before initializing runner.  Let me know if that is ok.

MikeThomsen commented:
@mans2singh Have you had a chance to work on any of these?

mans2singh commented:
Hi @MikeThomsen - Thanks for your review and comments.  

I will work on your and @joewitt 's recommendation this weekend.

mans2singh commented:
@joewitt @MikeThomsen - 

I've updated the code based on your comments.  

I've also added a docker gist [influxdb-compose.xml](https://gist.github.com/mans2singh/5ee90620314d4bcf26d6c65de2540c77#file-influxdb-compose-xml) based on docker compose xml mentioned by @MikeThomsen .  This docker compose starts a local influxdb and runs curl commands to create a `test` database and populate it with a metric every 5 seconds.  To launch it we need to run `docker-compose -f <file-name> up` on the local machine and then we can run the integration tests for the query processor or run Nifi locally with processor configured for scheduled query.

I am also including two templates ([flow file driven](https://gist.github.com/mans2singh/5ee90620314d4bcf26d6c65de2540c77#file-nifi-influxdb-flow-file-driven-template) and [timer driven](https://gist.github.com/mans2singh/5ee90620314d4bcf26d6c65de2540c77#file-nifi-influxdb-scheduled-query-template)) to assist in testing the InfluxDB query processor.

Please let me know if your comments/recommendations.

Thanks again.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/AbstractITInfluxDB.java:
```diff
@@ -36,14 +36,7 @@
-            result = influxDB.query(new Query("DROP database " + dbName, dbName));
-            Thread.sleep(1000);
-        }
+        cleanUpDatabase();
```
What I meant was that the cleanup functionality shouldn't be called in the init, but should be called from the `@After` annotated method.

mans2singh replied:
I call the cleanup in setup as a precaution so that if there is any conflicting/previously existing data in the test database, it is removed and does not fail the integration test which depend on number of rows inserted.  If you think it is unnecessary, I can remove it.

mans2singh commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -72,6 +74,16 @@
 
+    public static final PropertyDescriptor INFLUX_DB_SCHEDULED_QUERY = new PropertyDescriptor.Builder()
+            .name("influxdb-scheduled-query")
+            .displayName("InfluxDB Schedued Query")
```
Changed the attribute to INFLUX_DB_QUERY.  The description mentions that flow files and timed query both are allowed.

MikeThomsen commented:
Verified that both flows work against the Docker Compose configuration. Looking very good so far.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,258 @@
+
+            outgoingFlowFile = incomingFlowFile;
+        } else {
+            charset = Charset.forName(context.getProperty(CHARSET).evaluateAttributeExpressions().getValue());
```
These should be evaluated against the flowfile's attributes with `evaluateAttributeExpressions(FlowFile)`. I think it will work if the flowfile is null. A test against that wouldn't hurt just to be safe as well.

mans2singh replied:
Updated to use flow file for EL.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,258 @@
+            if ( result == null ) {
+                String message = "Query Result was " + result;
+                getLogger().error(message);
+                throw new NullPointerException(message);
```
I don't think this really works, and you said you were pretty sure it won't return a null if it fails. So I'm fine with ripping this out.

mans2singh replied:
Removed check for null.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,258 @@
+                session.transfer(outgoingFlowFile, REL_FAILURE);
+            }
+
+            session.getProvenanceReporter().send(outgoingFlowFile, makeProvenanceUrl(context, database),
```
You don't want to report provenance here if you have stuff sent to REL_FAILURE.

mans2singh replied:
Removed provenance for failure scenario.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,258 @@
+            final long endTimeMillis = System.currentTimeMillis();
+
+            if ( ! result.hasError() ) {
+                outgoingFlowFile = session.putAttribute(outgoingFlowFile, INFLUX_DB_EXECUTED_QUERY, String.valueOf(query));
```
You might want to put this on the failure scenario too because there could be syntax errors getting reported. This is especially going to be helpful if the user is dynamically generating the query.

mans2singh replied:
If there is any error the `hasError` returns true and the flow file is transferred to success.  Let me know if I've missed anything.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,258 @@
+                (endTimeMillis - startTimeMillis));
+        } catch (Exception exception) {
+            outgoingFlowFile = populateErrorAttributes(session, outgoingFlowFile, query, exception.getMessage());
+            if ( exception.getCause() instanceof SocketTimeoutException ) {
```
For readability, I think you should break this out into two catch blocks since you're handling one specific version very differently.

mans2singh replied:
The query does not throw any exception - I just added this so that if there is a socket timeout, the flow file is transferred to retry relation rather than failure.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    protected String makeProvenanceUrl(final ProcessContext context, String database) {
```
Adding a flowfile parameter here would help with the EL support.

mans2singh replied:
Added flow file parameter.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/AbstractITInfluxDB.java:
```diff
@@ -0,0 +1,84 @@
+
+    protected void initInfluxDB() throws InterruptedException, Exception {
+        influxDB = InfluxDBFactory.connect(dbUrl,user,password);
+        cleanUpDatabase();
```
Not needed.

mans2singh replied:
Removed as recommended.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,231 @@
+        assertEquals("Value should be equal", 1, flowFiles.size());
+        assertEquals("Value should be equal",null, flowFiles.get(0).getAttribute(ExecuteInfluxDBQuery.INFLUX_DB_ERROR_MESSAGE));
+        assertEquals("Value should be equal",query, flowFiles.get(0).getAttribute(ExecuteInfluxDBQuery.INFLUX_DB_EXECUTED_QUERY));
+        flowFiles.get(0).assertContentEquals(
```
If you can parse this with Gson and evaluate it that way, the test will be less likely to be brittle as InfluxDB evolves.

MikeThomsen replied:
I should also mention that there are Groovy JUnits in the codebase already, so if you want to redo these tests w/ Groovy following other examples that is acceptable. I bring that up mainly because Groovy is a lot cleaner than Java for handling JSON, Maps, etc.

mans2singh replied:
Updated to parse json and compare typed results.

mans2singh commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -0,0 +1,258 @@
+                return;
+            }
+
+            if ( incomingFlowFile.getSize() == 0) {
```
Changed to use query property if flow file is empty.

mans2singh commented:
@MikeThomsen - 

I've updated the code based on your comments.  Let me know if you have any more recommendations.

Thanks

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/main/java/org/apache/nifi/processors/influxdb/ExecuteInfluxDBQuery.java:
```diff
@@ -209,32 +204,40 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
-    protected String makeProvenanceUrl(final ProcessContext context, String database) {
+    protected String getQuery(final ProcessSession session, Charset charset, FlowFile incomingFlowFile)
+            throws IOException {
```
What I have done in a few processors is something like this:
```
String getQuery (final ProcessSession session, final ProcessContext context, Charset charset, FlowFile incomingFlowFile) throws IOException {
    if (context.getProperty(QUERY).isSet()) {
        return context.getProperty(QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
    } else {
        //Load from file
       if (StringUtil.isBlank(query)) {
           throw new IOException("Query was missing from the flowfile body");
       } else {
           return query;
       }
    }
}
```

MikeThomsen replied:
You don't have to redo it that way, it's just a suggestion.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITExecuteInfluxDBQuery.java:
```diff
@@ -110,8 +125,72 @@ public void testCreateDB() {
-        flowFiles.get(0).assertContentEquals("{\"results\":[{}]}");
+        QueryResult queryResult = gson.fromJson(new StringReader(new String(flowFiles.get(0).toByteArray())), QueryResult.class);
+        assertEquals("results array should be empty", 1, queryResult.getResults().size());
+        assertEquals("No series", null, queryResult.getResults().get(0).getSeries());
```
Let's use `assertNull` here to make it a bit more readable.

mans2singh replied:
Corrected.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITExecuteInfluxDBQuery.java:
```diff
@@ -110,8 +125,72 @@ public void testCreateDB() {
 
-        flowFiles.get(0).assertContentEquals("{\"results\":[{}]}");
+        QueryResult queryResult = gson.fromJson(new StringReader(new String(flowFiles.get(0).toByteArray())), QueryResult.class);
+        assertEquals("results array should be empty", 1, queryResult.getResults().size());
```
I think it would be a good addition to have: `assertNotNull(queryResult.getResults())` before this.

mans2singh replied:
Corrected.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITExecuteInfluxDBQuery.java:
```diff
@@ -110,8 +125,72 @@ public void testCreateDB() {
+        String query = "select * from water";
+        runner.setProperty(ExecuteInfluxDBQuery.INFLUX_DB_QUERY, query);
+
+        byte [] bytes = new byte [] {};
```
Redundant because you can use `runner.enqueue("")` for cases like this where you're not examining the original input flowfile contents.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/ITExecuteInfluxDBQuery.java:
```diff
@@ -110,8 +125,72 @@ public void testCreateDB() {
+
+        List<MockFlowFile> flowFiles = runner.getFlowFilesForRelationship(ExecuteInfluxDBQuery.REL_SUCCESS);
+        assertEquals("Value should be equal", 1, flowFiles.size());
+        assertEquals("Value should be equal",null, flowFiles.get(0).getAttribute(ExecuteInfluxDBQuery.INFLUX_DB_ERROR_MESSAGE));
```
`assertNull` is a better choice here.

mans2singh replied:
Corrected.

MikeThomsen commented on nifi-nar-bundles/nifi-influxdb-bundle/nifi-influxdb-processors/src/test/java/org/apache/nifi/processors/influxdb/TestExecutetInfluxDBQuery.java:
```diff
@@ -126,27 +131,28 @@ protected QueryResult executeQuery(ProcessContext context, String database, Stri
         byte [] bytes = "select * from /.*/".getBytes();
```
You can do `runner.enqueue("select * from /.*/")` if you want.

mans2singh commented:
@MikeThomsen - 

I have updated the test cases regarding usage of assertions.  Let me know if there is anything else outstanding.

Thanks

timhallinflux commented:
@joewitt and @MikeThomsen -- you can also just quickly spin up the TICK stack via the sandbox.
Have a look here... https://github.com/influxdata/sandbox

If @joewitt is no longer technical enough to run this, I'll be more than happy to do a screen share and help him out! :-)

Thank you @mans2singh for contributing this!

MikeThomsen commented:
@timhallinflux @mans2singh already provided a decent enough Docker Compose test for it. So unless @joewitt or someone else sees something that I missed that's a show stopper, I think we're good to go on merging this to master when 1.6 is released into the wild.

MikeThomsen commented:
Forgot to add that a link to a comprehensive sandbox is appreciated @timhallinflux

mans2singh commented:
Thanks @MikeThomsen for your reveiw/advice.  Also thanks @timhallinflux for the pointers on influxdb sandbox.

Please let me know if there is other feedback.

Mans

MikeThomsen commented:
@mans2singh I went to merge your changes this morning and found that you have some integration tests that are broken now with the changes made to EL support in the `PropertyDescriptor.Builder` class. It seems your breakage is limited mainly here:

```
protected String makeProvenanceUrl(final ProcessContext context, String database, FlowFile flowFile) {
    return new StringBuilder("influxdb://")
        .append(context.getProperty(INFLUX_DB_URL).evaluateAttributeExpressions(flowFile).getValue()).append("/")
        .append(database).toString();
}
```

You marked that property as `VARIABLE_REGISTRY` and it is failing because you're applying it flowfiles. The existing calls to `expressionLanguageSupported(boolean)` in the abstract class also need to be fixed.

So just do a rebase against the latest master and take care of those. If there are no regressions after your commit (ie your flows and tests run) IMO we're ready to merge.

mans2singh commented:
@MikeThomsen - 

I've updated the code (added expression language scope and updated tests) based your review comments.  The integration tests are passing.  

Please let me know if there is any other comment.

Thanks

Mans

MikeThomsen commented:
Approved and merged. Thanks @mans2singh

timhallinflux commented:
This is great!  Thank you Mike and Mans.  Mans.... if you reach out to me at tim at influxdata dot com
we would be happy to send you some stickers and a hoodie in appreciation for your efforts here.

MikeThomsen commented:
@timhallinflux if you don't mind, please take some time to think about additional NiFi-related use cases and either post them to the dev mailing list or add them as Jira tickets.

MikeThomsen commented:
@timhallinflux Based on my extremely limited knowledge of what InfluxDB can do, I don't know if even the U or D from CRUD are really applicable to InfluxDB for example. What I'd like to see is some input that would help us mold this bundle to maxing out the reasonable use cases where NiFi and InfluxDB can sensibly work together if that makes sense.

timhallinflux commented:
Yep.  Working on a blog post to highlight both parts...writer and reader.  

U is not really possible. For D, both DELETE and/or DROP statements are allowed through InfluxQL.

mans2singh commented:
@MikeThomsen  @joewitt - Thanks for your help in making this contribution possible.

@timhallinflux - Please let me know if you have any additional enhancements possible.  I will contact you 
via email.

Thanks again everyone.
